lib/rspec/matchers/built_in/base_matcher.rb

```diff
@@ -96,6 +99,15 @@ def self.matcher_name
         end
 
         # @private
+        def matcher_name
+          if defined?(@matcher_name)
+            @matcher_name
+          else
+            self.class.matcher_name
```

Comment:
Because `#matcher_name=` is set only when aliased (i.e. `@matcher_name` is nil when using matchers with their original name)..

The class method is defined [here](https://github.com/rspec/rspec-expectations/blob/41869a2abfbacede36391bd22f7257a5c840f7fd/lib/rspec/matchers/built_in/base_matcher.rb#L94-L96).